File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-6.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        heights.append(0)

        maxArea= 0

        for i,h in enumerate(heights):
                while stack and h < heights[stack[-1]]:
                        height = heights[stack[-1]]
                        stack.pop()

                        if not stack:
                                width = i
                        else :
                                width  = i - stack[-1]-1
                        maxArea = max(maxArea, height * width)
                stack.append(i)
        return maxArea

File: Data Structures & Algorithms/max-water-container/submission-6.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    maxArea(heights) {
        let left = 0;
        let right = heights.length-1;
        
        let maxArea = 0;

        while(left < right){
            let l = Math.min(heights[left], heights[right]);
            let w = right - left;
            let area = l * w;

            maxArea  = Math.max(area, maxArea);

            if(heights[left] < heights[right]){
                left++;
            }else{
                right --;
            }

        }
        return maxArea;
    }
}
